engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/EqualEvaluator.java

```diff
@@ -60,21 +60,29 @@ else if (left instanceof BigDecimal && right instanceof BigDecimal) {
         }
 
         else if (left instanceof Iterable && right instanceof Iterable) {
-            return CqlList.equal((Iterable<?>) left, (Iterable<?>) right);
+            return CqlList.equal((Iterable<?>) left, (Iterable<?>) right, context);
         }
 
         else if (left instanceof CqlType && right instanceof CqlType) {
             return ((CqlType) left).equal(right);
         }
 
-        return Context.getContext().objectEqual(left, right);
+        if (context != null) {
+            return context.objectEqual(left, right);
+        }
+
+        return null;
```

Comment:
I think this should throw, not return a silent null. If there is no context, it is an error to attempt to compare objects that require context to do so.

Reply:
Hmm... That set off a chain reaction where virtually every single NNNEvaluator needed to be updated. Also, it implies that the CqlTypes like Interval also need Context in order to be compared since their points could be types only comparable by a data provider. I think that's actually correct, but it means that any CqlType is tied to an instance of Context:

```
Context c1 = new Context();
Interval i1 = new Interval(1, true, 2, false, c1);

Context c2 = new Context();
Interval i2 = new Interval(1, true, 2, false, c2);
```
We currently can't create two contexts in the same thread because of the ThreadLocal, nor can we use the built-in aysnc work queues because the work may be allocated to a different thread than was used to create the context.

Anyway... Thoughts?

Reply:
Why couldn't they use the same context?

Reply:
I'm not saying intervals couldn't share context. Let me try that again:

```cql
Context c1 = new Context(qdmLibrary);
c1.registerDataProvider(qdmUrl, qdmProvider);
Interval i1 = new Interval(qdmType, true, qdmType, false);

Context c2 = new Context(fhirLibrary);
c2.registerDataProvider(fhirUrl, fhirProvider);
Interval i2 = new Interval(fhirType, true, fhirType, false);
```
For uses where you have two or more Contexts sharing a Thread, or cases where a Context switches to a different thread, the current design breaks because of the ThreadLocal storage. In the example above with the current design, the first interval will use the second context that's constructed. Ultimately, if you want to break the 1:1 Context:Thread requirement, you have to pass in the context you want to use to the Interval for it to correctly implement "Equals" and so on due to the fact that the point values can be non-CQL types.

```cql
Context c1 = new Context();
Interval i1 = new Interval(1, true, 2, false, c1); // Note the context being passed in
```

Imagine a scenario where you have hundreds or thousands of Contexts that need to be evaluated. You can have one thread queueing up the work (a producer thread) by creating contexts, and then a pool of worker threads that do the evaluation. You can't do such a thing currently because the producer thread can only have one context at a time (due to ThreadLocal) and the worker threads can't operate on Context created somewhere else (due to the ThreadLocal).

Does that help clarify?

Reply:
I see, you're just saying that to do this 100% correctly, all those places that are doing equality/equivalent determination will need to have a context passed in. I think we're okay to keep that as an edge case because all interval types currently definable in CQL use CQLTypes as their point types.

Reply:
In other words, yes, in theory non-CQLTypes can be point types, but there's no code path that will produce that behavior right now, at least not without quite a few other issues arising (like the fact that the point type wouldn't have minimum/maximum/successor/predecessor behavior defined).

Reply:
I added a check for CQL types in the constructor of the Interval so in the event some non-CQL type is created without Context it explodes.